### PR TITLE
criteria search implementation

### PR DESCRIPTION
**Pull request by c-kalenyk**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/repository/book/BookSpecificationProviderManager.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+@RequiredArgsConstructor
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException(
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/repository/book/BookSpecificationProviderManager.java]: change p to provider for clarity, replace exception with more precise and suitable NoSuchElementException

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 2]
```diff
@@ -35,6 +39,15 @@ public List<BookDto> findAll() {
         return bookMapper.toBookDtoList(bookRepository.findAll());
     }
 
+    @Override
+    public List<BookDto> search(BookSearchParameters searchParameters) {
+        Specification<Book> bookSpecification = bookSpecificationBuilder.build(searchParameters);
+        return bookRepository.findAll(bookSpecification)
+                .stream()
+                .map(bookMapper::toBookDto)
+                .toList();
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java]: use your toBookDtoList. You can rename this method to just toDto though

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/bookstore/repository/book/BookRepository.java [hunk 3]
```diff
@@ -1,9 +1,10 @@
-package mate.academy.bookstore.repository;
+package mate.academy.bookstore.repository.book;
 
 import mate.academy.bookstore.model.Book;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 import org.springframework.stereotype.Repository;
 
 @Repository
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/bookstore/repository/book/BookRepository.java]:
```suggestion
```

Comment (MentorMateAcademy) [reply src/main/java/mate/academy/bookstore/repository/book/BookRepository.java]: not mandatory

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

**Review by okuzan — APPROVED** [review]

**Pull request merged**